### PR TITLE
rbd-mirror: initial end-to-end test and associated bug fixes

### PR DESCRIPTION
**Pull request by dillaman**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by dillaman** [comment]
http://pulpito.ceph.com/jdillaman-2020-02-28_07:59:51-rbd-wip-jd-testing-distro-basic-smithi/

File: src/librbd/mirror/EnableRequest.cc [hunk 1]
```diff
@@ -65,7 +65,12 @@ void EnableRequest<I>::handle_get_mirror_image(int r) {
     r = cls_client::mirror_image_get_finish(&iter, &m_mirror_image);
   }
 
-  if (r == 0) {
+  if (r == 0 && m_mirror_image.state == cls::rbd::MIRROR_IMAGE_STATE_CREATING &&
+      !m_non_primary_global_image_id.empty()) {
+    // special case where rbd-mirror injects a disabled record to record the
+    // local image id prior to creating ther image
+    r = -ENOENT;
```
[the remark this answers is not in the document]
Comment (dillaman) [reply src/librbd/mirror/EnableRequest.cc]: Pushed a tweak

**Comment by dillaman** [comment]
> Also, it seems we could combine `rbd_mirror_journal.sh` and `rbd_mirror_snapshot.sh` into one

Agreed, it's on my todo list since I would also like to run the existing stress tests in both journal and snapshot-based mirroring mode.

**Comment by trociny** [comment]
@dillaman unittest_rbd_mirror crashed on jenkins. Is there any chance to get something useful from that run?

**Comment by dillaman** [comment]
Nothing is useful from Jenkins but I just recreated a failure in `TestMockImageReplayer.ReplayerRenamed` (which didn't fail in Jenkins).

**Comment by dillaman** [comment]
> /home/jdillaman/ceph/src/test/rbd_mirror/test_ImageReplayer.cc:482: Failure
> Value of: m_replayer->is_stopped()
>   Actual: false
> Expected: true
> /home/jdillaman/ceph/src/test/rbd_mirror/test_ImageReplayer.cc:872: Failure
> Expected equality of these values:
>   0
>   cond.wait()
>     Which is: -22
> /home/jdillaman/ceph/src/tools/rbd_mirror/ImageReplayer.cc: In function 'rbd::mirror::ImageReplayer<ImageCtxT>::~ImageReplayer() [with ImageCtxT = librbd::ImageCtx]' thread 7ff1f4d7a880 time 2020-03-02T14:10:16.717486-0500
> /home/jdillaman/ceph/src/tools/rbd_mirror/ImageReplayer.cc: 247: FAILED ceph_assert(m_state_builder == nullptr)
>  ceph version 15.1.0-1374-g50fb8f29b0 (50fb8f29b0e9d363b628924279551f169eafeb1a) octopus (rc)
>  1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x151) [0x7ff1f5cc2609]
>  2: (()+0x26f7f3) [0x7ff1f5cc27f3]
>  3: (()+0x796d02) [0x5599f96d6d02]
>  4: (rbd::mirror::ImageReplayer<librbd::ImageCtx>::~ImageReplayer()+0x9) [0x5599f96d6d19]
>  5: (rbd::mirror::TestImageReplayer<rbd::mirror::TestImageReplayerParams<(cls::rbd::MirrorImageMode)1> >::~TestImageReplayer()+0x63) [0x5599f9ad2913]
>  6: (rbd::mirror::TestImageReplayer_Resync_Test<rbd::mirror::TestImageReplayerParams<(cls::rbd::MirrorImageMode)1> >::~TestImageReplayer_Resync_Test()+0x17) [0x5599f9ad31d7]
>  7: (void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*)+0x4d) [0x5599f9a88b6d]
>  8: (testing::TestInfo::Run()+0xdd) [0x5599f9a7dc5d]
>  9: (testing::TestSuite::Run()+0xbd) [0x5599f9a7dd9d]
>  10: (testing::internal::UnitTestImpl::RunAllTests()+0x47c) [0x5599f9a7e2bc]
>  11: (bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)+0x4d) [0x5599f9a8908d]
>  12: (testing::UnitTest::Run()+0x9d) [0x5599f9a7e4dd]
>  13: (main()+0xc1) [0x5599f95a9961]
>  14: (__libc_start_main()+0xf3) [0x7ff1f51861a3]
>  15: (_start()+0x2e) [0x5599f95ced9e]
>

**Comment by dillaman** [comment]
OK, hopefully fixed the `TestMockImageReplayer.ReplayerRenamed` and `TestImageReplayer/1.Resync` test races.

**Pull request merged**